File: src/Test/StateMonadFFI.fs
module Test.StateMonadFFI
type StateResult<'s, 'a> = { value: 'a; state: 's }
type State<'s, 'a> = State of ('s -> StateResult<'s, 'a>)
let runState (State f) s = f s
let bind (State f) g = State (fun s -> let result = f s in runState (g result.value) result.state)
let pure' a = State (fun s -> { value = a; state = s })
let get = State (fun s -> { value = s; state = s })
let put s = State (fun _ -> { value = (); state = s })
let modify f = bind get (fun s -> put (f s))
let rec chain n = if n = 0 then pure' () else bind (modify (fun x -> x + 1)) (fun _ -> chain (n - 1))
let runStateMonadFFI (n: obj) =
    let rec loop i acc =
        if i = 0 then acc
        else
            let result = runState (chain (unbox<int> n)) 0
            loop (i - 1) (acc + result.state)
    loop 20 0 :> obj

File: src/Test/StateMonadFFI.erl
-module(test_stateMonadFFI@foreign).
-export([runStateMonadFFI/1]).
bind(M, F) -> fun(S) ->
  { A, S1 } = M(S),
  (F(A))(S1)
end.
pure(A) -> fun(S) -> { A, S } end.
get() -> fun(S) -> { S, S } end.
put(S) -> fun(_) -> { 0, S } end.
modify(F) -> bind(get(), fun(S) -> put(F(S)) end).
loop(0) -> pure(0);
loop(N) -> bind(modify(fun(S) -> S + 1 end), fun(_) -> loop(N - 1) end).
runStateMonadFFI(N) ->
  { _, FinalState } = (loop(N * 20))(0),
  FinalState.
